feat: add EzBarChart component (#958)
ezcater/recipe: packages/recipe/src/components/EzBarChart/EzBarChart.md

---
name: EzBarChart
title: Bar Chart
category: Data
path: '/components/ez-bar-chart'
tags: ['chart', 'graph', 'bar', 'line', 'progress']
---

> Bar charts are a [type of chart](<(https://help.flourish.studio/article/25-line-bar-and-pie-charts)>) consisting of a series of bars that describe a dataset.

Reviewer: This link doesn't work.
```suggestion
Bar charts are a [type of chart](https://help.flourish.studio/article/25-line-bar-and-pie-charts) consisting of a series of bars that describe a dataset.
```

---

## Best Practices

- Bar charts are used to show change over time, especially for a single series

> - Bar charts are good for comparing relative size

Reviewer: This is certainly fine for now, but do we want to reach out to UX content about what this should include for our design system?

Author: sure. i was trying to get this out quickly without another design loop since they needed it earlier this week

---

## Examples

### Basic bar chart

> Bar charts should include a title and container in an `EzCard` container, the title can optionally have a description and icon.

Reviewer: ```suggestion
Bar charts should include a title and be contained in an `EzCard`. The title can optionally have a description and icon.
```

> To chart a data set as a bar chart, provide a dataset as a `data` array of objects with `x` and `y` values, for example: `data = [{x: 'first', y: 1}, {x: 'second', y: 2}, {x: 'third', y: 3}]`

Reviewer: Perhaps this can be formatted to be easier to read?

---

To create a bar chart, provide a data set as a `data` array of objects with `x` and `y` values. For example:

```js
data = [
  {x: 'first', y: 1},
  {x: 'second', y: 2},
  {x: 'third', y: 3},
];
```

> Always provide `title` and a detailed `description` for assistive devices.

Reviewer: I would move this to an "Accessibility" section near the bottom to match our other component docs.

Author: it's part of the required props though?

Typically, the independent axis will be the x-axis and the dependent axis will be the y-axis. To format the point labels on each axis, optionally pass a function to each `dependentAxisLabelFormatter` and/or `independentAxisLabelFormatter`. To customize which labels appear on each axis, optionally pass an array of values of the same type as the respective `data` values to each `dependentAxisLabelValues` and/or `independentAxisLabelValues`.

> Bar colors can be customized with [theme colors](http://localhost:8000/guides/theming/#colors), the default is `primary`.

Reviewer: Should this be in a separate "Bar Chart Colors" section to match our other component doc pages?

```suggestion
Bar colors can be customized with [theme colors](/guides/theming/#colors). The default color is `primary`.
```

> If your dataset may sometimes be an empty array, work with design to render something other than a blank chart, perhaps an `EzBlankState`.

Reviewer: ```suggestion
If your data set may sometimes be an empty array, work with design to render something other than a blank chart, perhaps an [`EzBlankState`](/components/ez-blank-state).
```

```jsx
() => {
  const DATA = [
    {x: 'January', y: 100000},
    {x: 'February', y: 120000},
    {x: 'March', y: 145000},
    {x: 'April', y: 140000},
    {x: 'May', y: 180000},
    {x: 'June', y: 145000},
    {x: 'July', y: 200000},
    {x: 'August', y: 180000},
    {x: 'September', y: 200000},
    {x: 'October', y: 140000},
    {x: 'November', y: 170000},

>     {x: 'Decemeber', y: 180000},

Reviewer: ```suggestion
    {x: 'December', y: 180000},
```

Author: i fixed this in https://github.com/ezcater/recipe/pull/961

Reviewer: Could we fix it in this PR as well? I think it's better to keep reviews and review fixes to their respective PRs, then just rebase any PRs based on it after. It keeps everything clean IMO, but up to you.

Author: the bar chart PR was first, so line chart should be merged into it?

Reviewer: I mean we can do it this way for this set of PRs if you like, but in general I think it's better practice to go the other way. Merge PRs in the order they're opened and then rebase anything based on them after the merge. This means you can keep working without worry that something you're working on now that's based off older work must get merged first. I would not be able to do it that way with the Storybook Docs work for instance. It also keeps all the code and reviews clean and relevant for that particular PR, not spread across PRs. We can talk about it more offline if you want!

  });

  const independentAxisLabelFormatter = t => `${t.substring(0, 3)}`;
  const dependentAxisLabelFormatter = t => `${formatter.format(t)}`;

  const DEPENDENT_AXIS_LABEL_VALUES = [0, 50000, 100000, 150000, 200000, 250000, 300000];

  return (
    <EzPage>
      <EzCard title={METRIC} subtitle={TIME_PERIOD} actions={<EzHeading size="1">{SUM}</EzHeading>}>
        {!DATA || DATA.length === 0 ? (
          <EzBlankState
            title="No data available"
            description="We may not have enough information yet."
          />
        ) : (
          <EzBarChart
            data={DATA}
            description={description}
            dependentAxisLabelFormatter={dependentAxisLabelFormatter}
            dependentAxisLabelValues={DEPENDENT_AXIS_LABEL_VALUES}
            independentAxisLabelFormatter={independentAxisLabelFormatter}
            title={title}
          />
        )}
      </EzCard>
    </EzPage>
  );
};
```

---

## Props

```jsx-hide-controls
  <PropsTable propsData={[
    {
      name: 'barColor',
      types: ['EzThemeColors'],
      defaultValue: 'primary',
      description: 'Bar chart bar color.',
    },
    {
      name: 'data*',

>       types: ['array of datum'],

Reviewer: ```suggestion
      types: ['Datum[]'],
```

      description: 'Array of data with x and y values of type string, number, array of strings, or function.',
    },
    {
      name: 'description*',
      types: ['string'],
      description: 'Detailed accessible description for the chart.',

>     },

Reviewer: I would put any required props at the top, in alphabetical order.

Author: this is in alphabetical order, to match the props passed between the components

    {
      name: 'dependentAxisLabelFormatter',

>       types: ['function'],

Reviewer: For consistency with our other component docs.
```suggestion
      types: ['func'],
```

      description: 'The function to format the dependent axis points.',
    },
    {
      name: 'dependentAxisLabelValues',

>       types: ['array of axis label values'],

Reviewer: ```suggestion
      types: ['AxisLabelValue[]'],
```

      description: 'Array of axis label values to customize the dependent axis labels.',
    },
    {
      name: 'independentAxisLabelFormatter',

>       types: ['function'],

Reviewer: For consistency with our other component docs.
```suggestion
      types: ['func'],
```

      description: 'The function to format the independent axis points.',
    },
    {
      name: 'independentAxisLabelValues',

>       types: ['array of axis label values'],

Reviewer: ```suggestion
      types: ['AxisLabelValue[]'],
```

      description: 'Array of axis label values to customize the independent axis labels.',
    },
    {
      name: 'title*',
      types: ['string'],
      description: 'Accessible title for the chart.',

>     },

Reviewer: Place required props at the top, in alphabetical order.

Author: this is in alphabetical order, to match the props passed between the components

  ]}>
  </PropsTable>
```

---

## Related components

> - [Progress](/components/ez-progress)

Reviewer: ```suggestion
- [EzProgress](/components/ez-progress)
- [EzBlankState](/components/ez-blank-state)
```
